BIP451: Dust UTXO Disposal Protocol (#2150)
bitcoin/bips: bip-XXXX.md

```
  BIP: ?
  Layer: Applications
  Title: Dust UTXO Disposal Protocol
  Authors: bubb1es <bubb1es71@proton.me>
           harris <imtux2@proton.me>
  Status: Draft
  Type: Specification
  Assigned: ?
  License: CC0-1.0
  Discussion: 2026-01-25: https://delvingbitcoin.org/t/disposing-of-dust-attack-utxos/2215
              2026-03-30: https://groups.google.com/g/bitcoindev/c/pr1z3_j8vTc
  Version: 0.1.0
```

## Abstract

This BIP specifies a standardized protocol for safely disposing of dust UTXOs by spending them to an OP_RETURN output with the entire value going to transaction fees. The protocol enables wallet software to remove unwanted small-value UTXOs, particularly those received in dust attacks, without creating new address linkages or degrading user privacy. The specification includes transaction format requirements, signature conventions enabling third-party batching, and validation rules for compliant implementations.

## Motivation

### The Dust Attack Problem

> Dust attacks are a well-documented privacy threat where attackers send small amounts of bitcoin to numerous addresses. When wallet software later consolidates these dust UTXOs with non-dust UTXOs, the attacker can analyze the blockchain to link previously unassociated addresses, potentially deanonymizing users.

Reviewer: Perhaps:

```suggestion
Dust attacks are a well-documented privacy threat where attackers send small amounts of bitcoin to numerous addresses. When wallet software later consolidates these dust UTXOs with non-dust UTXOs, the attacker use the [common-input-ownership heuristic](https://en.bitcoin.it/wiki/Privacy#Common-input-ownership_heuristic) to link previously unassociated addresses together and with future spending, reducing the privacy of users.
```

- **No New UTXOs**: OP_RETURN outputs are provably unspendable and not stored in the UTXO set.
- **No Address Linking**: Without a change output, there is no new address to link.
- **Permanent Removal**: The dust UTXOs are removed from the user's wallet entirely.
- **Miner Compensation**: OP_RETURN outputs are small, providing higher transaction fee rates.
- **No Cost to Victims**: Dust attack UTXO values are used to pay for their own disposal.

### Why Standardization

A standardized protocol enables:

- **Wallet Anonymity**: Transactions with a standard format cannot be used to fingerprint the wallet software a user is running.
- **Third-Party Batching**: Multiple dust disposals can be combined into single transactions, reducing overall block space consumption.
- **Best Practice Codification**: Ensures implementations follow privacy-preserving best practices.
- **Easy Identification**: Chain analysis tools can use disposal transactions to help trace the sources of dust attacks.

## Specification

### Transaction Format

A compliant dust disposal transaction MUST satisfy all the following requirements:

#### Overall

1. The nLockTime MUST be set to block height 0.

> 2. The fee rate MUST be at least 0.1 sat/vB.

Reviewer: Perhaps:

```suggestion
2. The fee rate MUST be at least the minimum transaction relay feerate (0.1 sat/vB at the time of writing).
```

2. All inputs MUST use the signature hash type `ALL|ANYONECANPAY` (0x81).
3. For Taproot (P2TR) inputs using key-path spending, implementations MUST explicitly append the signature hash type byte `ALL|ANYONECANPAY` (0x81) to enable ANYONECANPAY semantics, as the default sighash for Taproot (SIGHASH_DEFAULT, which omits the byte) does not include ANYONECANPAY.
4. All inputs must be confirmed in the blockchain at least one block deep.

#### Transaction Size

1. The transaction base size MUST be at least 65 bytes to meet the current bitcoin network minimum relay size standardness rule.
2. If smaller than 65 bytes, the output OP_RETURN value "ash" as ASCII (UTF-8) bytes (0x61, 0x73, 0x68) MUST be used to pad the transaction's size to 65 bytes.

#### Fees

1. The entire input value MUST go to fees (the single OP_RETURN output amount is zero).
2. The transaction fee rate MUST meet the bitcoin network minimum relay requirements (0.1 sat/vB for Bitcoin Core 28.3, 29.1, 30.0+).
3. The transaction fee rate MAY be higher based on the available dust UTXO amounts and transaction size.

### Dust Selection

Implementations SHOULD allow users to configure how dust UTXOs are selected based on:

> 1. The current and anticipated fee rates.

Reviewer: Since the feerate directly follows from the potential dust UTXO (i.e. its amount and output type), it is not obvious to me why the current or anticipated feerates play a relevant role: whenever the transaction would be created it would have the same feerate anyway.

I do see that sending a dust disposal transaction when the current demand is low would lead to quick confirmation which would quickly resolve the wallet’s dust situation. However, broadcasting at high feerates would give more time for multiple dust disposals to be aggregated. So, really any time seems fine for different reasons.

I would also point out that if wallets have opinionated default behavior on when to create and broadcast dust disposal transactions, the broadcast time might become a (weak?) wallet fingerprint.

Author: We were thinking here about the dynamic `mempoolminfee` and that if mempools are full or near full users (or their wallets) may have to make a judgement call when selecting viable dust UTXOs to try to dispose of.  

Would it make more sense to rewrite this section to make the only user configurable parameters the max UTXO amount that is considered dust. This is needed since we don't know how the amounts may change in the future. And otherwise dust UTXOs should always be selected if the fee rate of their dispose tx is at least the `mempoolminfee` rate?

Reviewer: A configurable max amount sounds like a good idea generally. Broadcasting transactions only when they make it into the mempool also sounds reasonable. Although it is not logistically an issue if the transaction were created when it doesn’t propagate. If you e.g., create a transaction in Bitcoin Core it will be rebroadcast every 12–24h until it gets confirmed. Especially if you intend to use `privatebroadcast` it doesn’t seem like a big privacy issue either, if it were submitted to peers a few more times before it propagates widely.

Author: I forgot that Bitcoin Core will keep rebroadcasting so I think it's safe here to just focus on letting the user configure the max amount.

Batch dust disposal transactions must follow all other transaction construction requirements for non-batched dust disposal transactions.

This mempool-based approach preserves user privacy while enabling efficient batching:

1. If no unconfirmed dust disposal transactions exist in the mempool, broadcast a new dust disposal transaction to the network.
2. When one or more unrelated, unconfirmed dust disposal transactions are found, try to create a compliant batch transaction with them.
3. If no valid batch transaction can be created, fall-back to option 1.

### Security Considerations

#### Transaction Signing

- **Key Security**: Signing dust disposal transactions requires signing with the user's wallet private keys. This could be a risk for cold storage wallets where the key or keys needed to sign are not easily accessible.
- **Transaction Correctness**: Transaction signers must carefully review and verify that only dust UTXOs are spent and no other inputs are signed.

#### Privacy Preservation

> - **Network surveillance**: Internet service providers and other internet monitors may be able to determine the nodes that initially broadcast a dust disposal transaction. If available the `sendrawtransaction -privatebroadcast` RPC feature should be used (Bitcoin Core 31.0).

Reviewer: Just the version is a bit underspecified, and it may become available in other software later. How about:

```suggestion
- **Network surveillance**: Internet service providers and other internet monitors may be able to determine the nodes that initially broadcast a dust disposal transaction. If available the `sendrawtransaction -privatebroadcast` RPC feature should be used (e.g., available in Bitcoin Core 31.0+).
```

- **Targeted dust**: An attacker may send dust UTXOs to a victim's wallet specifically to trick them into disposing of it and revealing the address's public key. This targeting could happen in anticipation of any future cryptographic attack that requires an exposed public key.

## Rationale

### Why Empty or "ash" OP_RETURN Data?

- **Minimal Size**: Empty data (2 bytes: OP_RETURN OP_0) minimizes the transaction size.
- **Standardization**: Consistent transaction construction eliminates wallet fingerprinting.
- **Padding Option**: The "ash" string (5 bytes: OP_RETURN OP_PUSHBYTES_3 "ash") provides a standardized way to meet the minimum transaction size; e.g., for a single P2TR dust input.
- **Semantic Meaning**: The word "ash" metaphorically represents the result of "burning" the dust.

### Why Per-Address Transactions?

Consolidating dust from multiple addresses for the same wallet creates the same privacy harm that dust attacks attempt to achieve. By requiring implementations to create separate transactions per address (by default), the protocol ensures dust disposal doesn't harm privacy.

> ### Why 65 Byte Minimum?

Reviewer: ```suggestion
### Why 65-Byte Minimum?
```

Bitcoin relay policy enforces a minimum transaction base size of 65 bytes to prevent certain attack vectors. Compliant transactions must meet this threshold to be relayed by standard nodes.

### Why 0.1 sat/vB Minimum Fee Rate?

> Dust disposal transactions must meet the minimum relay fee rate, which is currently 0.1 sat/vB (Bitcoin Core 28.3, 29.1, 30.0+). This allows dust UTXOs to be disposed of economically even when their value is small. Implementations targeting older node versions may need higher minimum fee rates.

Reviewer 2: Maybe address theoretical updating if the minimum relay fee rate changes in the future.

Author: Would adding a sentence like this work? 

"Dust disposal transactions should always pay at least the current standard minimum relay fee rate if it changes in the future."

### List dust

1. Create payment addresses for multiple address types, send dust and non-dust amounts to these addresses, verify that listing dust only returns UTXOs at or below the configured dust threshold (e.g. 1000 sats).
2. Create confirmed and unconfirmed dust UTXOs in the test wallet, verify listing dust only returns the confirmed dust UTXOs.

### Spending dust

All valid dust disposal transactions should be verified to be accepted into the bitcoind (Bitcoin Core 30.0+) mempool.

1. Spending a single witness (Bech32m/P2TR) dust UTXO must produce a dust disposal transaction with a single "ash" OP_RETURN output.
2. Spending multiple dust UTXOs always produces a single empty OP_RETURN output regardless of script type.
3. Spending a single 2-of-2 P2SH multisig dust UTXO produces a single empty OP_RETURN output.
4. All dust UTXOs sent to the same address are disposed of in the same transaction.
5. Dust disposal transactions only include confirmed dust UTXOs.

#### Example dust disposal transaction sizes

|                   | P2PKH | P2SH (2-3) | P2WPKH | P2WSH (2-3) | P2TR  |
|-------------------|-------|------------|--------|-------------|-------|

> | Overhead (b)      | 10    | 10         | 10     | 10          | 10    | 

Reviewer: Nit: byte is abbreviated with a capital B
